bundles/org.openhab.binding.ecovacs/src/main/java/org/openhab/binding/ecovacs/internal/api/impl/EcovacsXmppDevice.java

```diff
@@ -0,0 +1,463 @@
+                oldFuture.cancel(true);
+            }
+            if (started) {
+                this.nextPing = scheduler.schedule(this::sendPing, INTERVAL - delta, TimeUnit.SECONDS);
+            }
+        }
+    }
+
+    private class IncomingMessageHandler extends AbstractIqRequestHandler {
+        private final EventListener listener;
+        private final ReportParser parser;
+        private final ConcurrentHashMap<String, CommandResponseHolder> pendingCommands = new ConcurrentHashMap<>();
+        private boolean disposed;
+
+        IncomingMessageHandler(EventListener listener) {
+            super("query", "com:ctl", Type.set, Mode.async);
+            this.listener = listener;
+            this.parser = new XmlReportParser(EcovacsXmppDevice.this, listener, gson);
+        }
+
+        void registerPendingCommand(String id, CommandResponseHolder responseHolder) {
+            pendingCommands.put(id, responseHolder);
+        }
+
+        void unregisterPendingCommand(String id) {
+            pendingCommands.remove(id);
+        }
+
+        void dispose() {
+            disposed = true;
+        }
+
+        @Override
+        public @Nullable IQ handleIQRequest(@Nullable IQ iqRequest) {
+            if (disposed) {
+                return null;
+            }
+
+            if (iqRequest instanceof DeviceCommandIQ) {
+                DeviceCommandIQ iq = (DeviceCommandIQ) iqRequest;
+
+                try {
+                    if (!iq.id.isEmpty()) {
+                        CommandResponseHolder responseHolder = pendingCommands.remove(iq.id);
+                        if (responseHolder != null) {
+                            synchronized (responseHolder) {
+                                responseHolder.response = iq.payload;
+                                responseHolder.notifyAll();
+                            }
+                        }
+                    } else {
+                        Optional<String> eventNameOpt = XPathUtils.getFirstXPathMatchOpt(iq.payload, "//ctl/@td")
+                                .map(n -> n.getNodeValue());
+                        if (eventNameOpt.isPresent()) {
+                            logger.trace("{}: Received event message XML {}", getSerialNumber(), iq.payload);
+                            parser.handleMessage(eventNameOpt.get(), iq.payload);
+                        } else {
+                            logger.debug("{}: Got unexpected XML payload {}", getSerialNumber(), iq.payload);
+                        }
+                    }
+                } catch (Exception e) {
+                    listener.onEventStreamFailure(EcovacsXmppDevice.this, e);
+                }
+            } else if (iqRequest instanceof ErrorIQ) {
+                StanzaError error = ((ErrorIQ) iqRequest).getError();
+                logger.trace("{}: Got error response {}", getSerialNumber(), error);
+                listener.onEventStreamFailure(EcovacsXmppDevice.this,
```

Comment:
Hmm, do you handle IQ responses in an IQ *request* handler? If so, then you may want to reconsider the implementation. Smack does provide high level API for sending IQ requests and gathering responses, including handling response timeouts.

Reply:
I know, but I can't use it due to Ecovacs' braindead implementation. A sample request-response pair looks like this:
```
2022-02-03 07:43:51.737 [TRACE] [.internal.api.impl.EcovacsXmppDevice] - XXX: SENT (0): <iq to='XXX@115.ecorobot.net/atom' id='GWpLy-18' type='set'><query xmlns='com:ctl'><ctl id="55792281" name="FW" td="GetVersion"/></query></iq>
2022-02-03 07:43:51.812 [TRACE] [.internal.api.impl.EcovacsXmppDevice] - XXX: RECV (0): <iq to="XXX@ecouser.net/e7a4f07a" type="set" id="099825546960" from="XXX@115.ecorobot.net/atom"><query xmlns="com:ctl"><ctl id="55792281" ret="ok">
```
You can see that they do not use the IQ ID field the way it's intended and return a different ID (with whatever source/meaning). Instead, the client (here: the binding) is supposed to add an ID field to the payload, and that one will be echoed to be able to match request and response. I tried to use `AbstractXMPPConnection.sendIqRequestAndWaitForResponse()` first ;-)
Additionally their response is a `set` request instead of a `result`. Out-of-bounds messages are sent as `set` requests as well.

Reply:
A little piece of me just died.

Reply:
Heh, I had precisely the same thought when seeing the first log with that communication. Nothing we can change here I guess, though...